cyberbotics/webots
# Drastically reduce package size

Following up #2591, this PR introduces Webots URLs. Webots URLs can be used in a world file to specify textures, meshes, sounds, etc. in a similar way as the https URLs introduced in #2591. Webots URLs come as an alternative to relative or absolute file references and identify a unique file in the Webots package.
A Webots URL has the following form: `webots://` following by a file path relative to `WEBOTS_HOME`.
For example: [webots://projects/samples/geometries/worlds/textures/webots_box.png](https://raw.githubusercontent.com/cyberbotics/webots/R2021a/projects/samples/geometries/worlds/textures/water_flow.jpg)

When releasing a new webots package, all world and proto files will be processed and all `webots://` URLs will be replaced with their equivalent `https://raw.githubusercontent.com/cyberbotics/webots/` URL, so that all the resources files (textures, meshes, sounds, etc.) can be omitted from the distributed package. This will significantly reduce the size of the Webots packages.

Released version will use the GitHub tag to refer to resources, e.g., R2021a, R2021a-rev1, R2021b, etc.
Nightly builds cannot use the GitHub tag for nightly builds because these tags are destroyed after 3 days. So, instead, they should use the commit sha.

- [x] Implement support for `webots://` URLs in Webots (useful only for Webots developers who checked out the webots repo).
- [ ] ~Add a "Download all resources for offline use" menu item (will download all resources in the cache of Webots).~
- [x] Replace the file URLs with `webots://` in all world files.
- [x] Replace the file URLs with `webots://` in all proto files.
    The following protos were converted manually as the script failed because of LUA code:
    - SimpleBuilding
    - PublicToilet
    - SimpleTree
    The following protos were skipped as the script failed because of LUA code, but no modifications was needed:
    - DirectionalPanel
    - ProtoMismatchFieldType (tests folder)
- [x] Replace the file URLs also in the webots_ros submodule. See https://github.com/cyberbotics/webots_ros/pull/59
- [x] Replace the file URLs also in the webots_ros2 repo. See https://github.com/cyberbotics/webots_ros2/pull/233
- [x] Replace the file URLs also in the webots-project. See https://github.com/cyberbotics/webots-projects/pull/4
- [x] Change the export of .x3d files to use `webots://` urls when possible.
- [x] Adapt webotsJS to support `webots://`
- [x] Replace the `webots://` of all world and proto files in the distribution.
- [x] Remove the resources files from the distribution.
- [x] Measure the resulting size of the new webots packages: on Windows, the new package is 0.99 GB instead of 1.64 GB.


Other ideas to reduce the package size:

- on Windows, remove the mingw32 binaries.
- remove the offline documentation, replace it with online documentation only.
- remove the e-puck cross-compiler &rarr; move it to a different repo.
- use EXTERNPROTO to move proto files outside of the distribution.
- move projects (worlds, protos, controllers) outside of the distribution.
